gdraw/gtextfield.c

```diff
@@ -1754,14 +1755,15 @@ return( true );
 
 	if ( gt->drag_and_drop ) {
 	    refresh = GTextFieldDoDrop(gt,event,end-gt->text);
+	    // curlen may be inaccurate now, but we recalculate after this guard set.
 	} else if ( gt->linesel ) {
-	    int j, e;
-	    gt->sel_start = gt->lines[i]; gt->sel_end = gt->lines[i+1];
-	    if ( gt->sel_end==-1 ) gt->sel_end = u_strlen(gt->text);
-	    for ( j=0; gt->lines[i+1]!=-1 && gt->sel_base>=gt->lines[i+1]; ++j );
-	    if ( gt->sel_start<gt->lines[i] ) gt->sel_start = gt->lines[i];
-	    e = gt->lines[j+1]==-1 ? u_strlen(gt->text): gt->lines[j+1];
-	    if ( e>gt->sel_end ) gt->sel_end = e;
+	    int j;
+	    for ( j=i; j>0 && gt->text[gt->lines[j]-1] != '\n'; --j )
```

Comment:
I'm having a bit of trouble reproducing the original issue; anyway, would you mind giving a brief description of how this is hit/what this does?

Reply:
I don't know exactly what the original code was supposed to do. Note that the old loop with `j` doesn't modify or test `j`. But conceptually what the code is supposed to do is clear: `linesel` is set to true on a triple-click and is supposed to select the line on which the click occurred. The intent of this new code is to select a "semantic" line -- a region of text between newlines -- rather than a visual line created by wrapping. There's no real reason to select the latter.

Reply:
The issue, based on the symptoms, was that the guard on the loop was usually false, but when it was true you'd get a freeze for the self-evident reason.

Reply:
I just realized I could be a bit more helpful on this. 

`gt->lines[]` is an array of offsets into `gt->text` marking the start of a new *visual* line. It's a datastructure maintained in other code as contents are added and the window dimension changed.